src/words/TokTok.jl

```diff
@@ -0,0 +1,245 @@
+        "\ufdfc", "\ufe69", "\uff04", "\uffe0", "\uffe1", "\uffe5", "\uffe6", "\u09fb",)
+
+# Left/Right strip, i.e. remove heading/trailing spaces.
+const LSTRIP = (" ",) => ""           
+const RSTRIP = ("\r", "\n", "\t", "\f", "\v",) => "\n"  
+# Merge multiple spaces.
+const ONE_SPACE = ("  ",) => " "
+
+const rules_atoms = Tuple(Iterators.flatten([
+        CURRENCY_SYM,
+        FUNKY_PUNCT_1,
+        FUNKY_PUNCT_2,
+        EN_EM_DASHES,
+        PROB_SINGLE_QUOTES,
+        OPEN_PUNCT,
+        CLOSE_PUNCT    
+    ])
+
+const rules_replaces = vcat(
+        NON_BREAKING,
+        AMPERCENT,
+        TAB,
+        PIPE,
+        STUPID_QUOTES_1,
+        STUPID_QUOTES_2,
+        ONE_SPACE
+    )
+
+const rules_splitting = vcat(LSTRIP, RSTRIP)
+
+"""
+    totok_tokenize(instring::AstractString)
+This tokenizer is a simple, general tokenizer, where the input has one sentence per line; thus only final period is tokenized. 
+Tok-tok has been tested on and gives reasonably good results for English, Persian, Russian, Czech, French, German, Vietnamese,
+Tajik, and a few others.
+"""
+function toktok_tokenize(instring::AbstractString)
+    ts = TokenBuffer(instring)
+    isempty(ts.input) && return ts.tokens
+ 
+    flag = handle_final_periods(ts)
+    
+    while !isdone(ts) && ts.idx <= flag
+        url_handler1(ts) ||   # these url handlers have priority over others
+        url_handler2(ts) ||
+        url_handler3(ts) || 
+        url_handler4(ts) ||
+        atoms(ts, rules_atoms) ||
+        replaces(ts, rules_replaces) || 
+        replaces(ts, rules_splitting, boundary = true) ||
+        repeated_character_seq(ts, ",", 2) || 
+        repeated_character_seq(ts, "-", 2) || 
+        repeated_character_seq(ts, ".", 2) ||
+        number(ts) ||
+        spaces(ts) || 
+        character(ts)
+    end
+    flush!(ts)
+    return ts.tokens
+end
+
+"""
+    handle_final_periods(::TokenBuffer)
+Handles the following rules using TokenBuffer API:  
+FINAL_PERIOD_1
+FINAL_PERIOD_2
+"""
+function handle_final_periods(ts::TokenBuffer)
+    flag = length(ts.input)
+    # handles FINAL_PERIOD_1 = r"(?<!\.)\.$"
+    if length(ts.input) >= 2 && ts.input[end] == "." && ts.input[end-1] != "."
+        flush!(ts, ".")
+        flag -= 1
+    end
+
+    # handles FINAL_PERIOD_2 = r"(?<!\.)\.\s*(["'’»›”]) *$"
+    if ts.input[end] in string.(Tuple("“”‘’›"))
+        i = flag - 1
+        while ts.input[i] in RSTRIP[1]
+            i -= 1
+        end
+        if ts.input[i] == "."
+            flag = i - 1
+            flush!(ts, ts.input[i:end])
+        end
+    end
+    return flag
+end
+
+"""
+    url_handler1(::TokenBuffer)
+Handles this rule from python using TokenBuffer API:  
+URL_FOE_1 = re.compile(r':(?!//)'), r' : '  
+"""
+function url_handler1(ts::TokenBuffer)
+    str = ":"
+    pattern = "//"
+    if ts.idx + length(pattern) <= length(ts.input) && ts.input[ts.idx] == str && 
```

Comment:
have  single outer 
```
if ts.idx + length(pattern) <= length(ts.input) && ts.input[ts.idx] == str
```
then inside that have 
```
if ts.input[ts.idx + 1 : ts.idx + length(pattern)] != pattern
...
else
    for ...
```

Similar applies in `url_handler2`